rasterio/_io.pyx

```diff
@@ -65,474 +65,124 @@ cdef bint in_dtype_range(value, dtype):
         99: np.finfo,
         102: np.finfo,
         105: np.iinfo,
-        117: np.iinfo
-    }
+        117: np.iinfo}
+
     key = np.dtype(dtype).kind
     if np.isnan(value):
         return key in ('c', 'f', 99, 102)
 
     rng = infos[key](dtype)
     return rng.min <= value <= rng.max
 
-# Single band IO functions.
-
-cdef int io_ubyte(
-        GDALRasterBandH band,
-        int mode,
-        int xoff,
-        int yoff,
-        int width,
-        int height,
-        np.uint8_t[:, :] buffer):
-    with nogil:
-        return GDALRasterIO(
-            band, mode, xoff, yoff, width, height,
-            &buffer[0, 0], buffer.shape[1], buffer.shape[0], 1, 0, 0)
-
-cdef int io_uint16(
-        GDALRasterBandH band,
-        int mode,
-        int xoff,
-        int yoff,
-        int width,
-        int height,
-        np.uint16_t[:, :] buffer):
-    with nogil:
-        return GDALRasterIO(
-            band, mode, xoff, yoff, width, height,
-            &buffer[0, 0], buffer.shape[1], buffer.shape[0], 2, 0, 0)
 
-cdef int io_int16(
+cdef int io_band(
         GDALRasterBandH band,
         int mode,
         int xoff,
         int yoff,
         int width,
         int height,
-        np.int16_t[:, :] buffer):
-    with nogil:
-        return GDALRasterIO(
-            band, mode, xoff, yoff, width, height,
-            &buffer[0, 0], buffer.shape[1], buffer.shape[0], 3, 0, 0)
+        object data):
+    """Read or write a region of data for the band.
 
-cdef int io_uint32(
-        GDALRasterBandH band,
-        int mode,
-        int xoff,
-        int yoff,
-        int width,
-        int height,
-        np.uint32_t[:, :] buffer):
-    with nogil:
-        return GDALRasterIO(
-            band, mode, xoff, yoff, width, height,
-            &buffer[0, 0], buffer.shape[1], buffer.shape[0], 4, 0, 0)
+    Implicit are
 
-cdef int io_int32(
-        GDALRasterBandH band,
-        int mode,
-        int xoff,
-        int yoff,
-        int width,
-        int height,
-        np.int32_t[:, :] buffer):
-    with nogil:
-        return GDALRasterIO(
-            band, mode, xoff, yoff, width, height,
-            &buffer[0, 0], buffer.shape[1], buffer.shape[0], 5, 0, 0)
+    1) data type conversion if the dtype of `data` and `band` differ.
+    2) decimation if `data` and `band` shapes differ.
 
-cdef int io_float32(
-        GDALRasterBandH band,
-        int mode,
-        int xoff,
-        int yoff,
-        int width,
-        int height,
-        np.float32_t[:, :] buffer):
-    with nogil:
-        return GDALRasterIO(
-            band, mode, xoff, yoff, width, height,
-            &buffer[0, 0], buffer.shape[1], buffer.shape[0], 6, 0, 0)
+    The striding of `data` is passed to GDAL so that it can navigate
+    the layout of ndarray views.
+    """
+    # GDAL handles all the buffering indexing, so a typed memoryview,
+    # as in previous versions, isn't needed.
+    cdef void *buf = <void *>np.PyArray_DATA(data)
```

Comment:
From the [numpy API docs]( http://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_DATA)

> If you have not guaranteed a contiguous and/or aligned array then be sure you understand how to access the data in the array to avoid memory and/or alignment problems.

So is the assumption that GDAL (with the addition of the `bufpixelspace` and `buflinespace` arguments) fully understands how to access the data, regardless of how it is laid out in memory?